File: apps/mo1-ios/ios/Sources/Mo1Core/Networking/TokenStore.swift
// TokenStore — Keychain abstraction (design §1 "Bearer トークン・Keychain 保管",
// mirrors token-store.ts). The production app backs this with the iOS Keychain
// (see KeychainTokenStore); the in-memory impl is used by tests. Logout MUST
// clear everything (design §3).
import Foundation

public struct StoredSession: Equatable, Sendable {
    /// Single opaque bearer token (theme8).
    public var token: String
    /// epoch-ms session expiry mirror (auth.SessionInfo.sessionExpiresAt).
    public var sessionExpiresAt: EpochMs
    public init(token: String, sessionExpiresAt: EpochMs) {
        self.token = token; self.sessionExpiresAt = sessionExpiresAt
    }
}

public protocol TokenStore: AnyObject, Sendable {
    func read() -> StoredSession?
    func write(_ session: StoredSession)
    func clear()
}

/// In-memory reference impl for tests (thread-safe via a lock).
public final class InMemoryTokenStore: TokenStore, @unchecked Sendable {
    private var session: StoredSession?
    private let lock = NSLock()
    public init() {}

    public func read() -> StoredSession? {
        lock.lock(); defer { lock.unlock() }
        return session
    }
    public func write(_ session: StoredSession) {
        lock.lock(); defer { lock.unlock() }
        self.session = session
    }
    public func clear() {
        lock.lock(); defer { lock.unlock() }
        session = nil
    }
}

File: apps/mo1-ios/src/token-store.ts
// token-store — Keychain abstraction (design §1 "Bearer トークン・Keychain 保管").
// The Swift app backs this with the iOS Keychain; the reference layer ships an
// in-memory impl for tests. Logout MUST clear everything (design §3).
export interface StoredSession {
  /** Single opaque bearer token (theme8). */
  token: string;
  /** epoch-ms session expiry mirror (auth.SessionInfo.sessionExpiresAt). */
  sessionExpiresAt: number;
}

export interface TokenStore {
  read(): StoredSession | null;
  write(session: StoredSession): void;
  clear(): void;
}

export class InMemoryTokenStore implements TokenStore {
  #session: StoredSession | null = null;

  read(): StoredSession | null {
    return this.#session;
  }

  write(session: StoredSession): void {
    this.#session = session;
  }

  clear(): void {
    this.#session = null;
  }
}
